src/lib.rs

```diff
@@ -336,22 +340,97 @@ impl Config {
+
+            let mut input_mtime = get_mtime(&file);
+            let output_mtime = get_mtime(&obj);
+
+            let mut cmd = compiler.to_command();
+            if msvc {
+                cmd.arg("/showIncludes").arg("/E");
+            } else {
+                cmd.arg("-M").arg("-MG");
+            }
+            cmd.arg(&file);
+            println!("running: {:?}", cmd);
+
+            let cpp_output = match cmd.output() {
+                Ok(ref out) if out.status.success() => {
+                    String::from_utf8(if msvc {
+                        out.stderr.clone()
+                    } else {
+                        out.stdout.clone()
+                    }).ok()
+                },
+                _ => {
+                    println!("warning: dependency check failed, building anyway: {:?}", file);
+                    input_mtime = output_mtime;
+                    None
+                }
+            };
+
+            if let Some(cpp) = cpp_output {
+                let files = if msvc {
+                    // Output is indented dependency per line, with "Note: ..." prefix 
+                    cpp
+                        .replace("Note: including file:", "")
+                } else {
+                    // Output is make rule with line continuations and multiple files per line
+                    cpp
+                        .replace("\\\n", "") // remove line continuations
+                        .replace("\\ ", "\0") // protect escaped spaces
+                        .replace(" ", "\n") // spaces to new lines
+                        .replace("\0", "\\ ") // bring escaped spaces back
+                };
+
+                input_mtime = files
+                    .lines().skip(1)
+                    .chain(vec![file.to_str().unwrap_or("")])
+                    .filter_map(|s| PathBuf::from(s.trim()).canonicalize().ok())
+                    .map(|f| dependencies
+                        .entry(f.clone())
+                        .or_insert(get_mtime(f))
+                        .clone())
+                    .max()
+                    .unwrap_or(FileTime::zero());
+            }
+
+            // build if a dependency is fresher than the output object
+            if input_mtime >= output_mtime {
+                println!("{:?} or one of its dependencies is {}s fresher than {:?}",
+                    file,
+                    input_mtime.seconds() - output_mtime.seconds(),
+                    obj);
+                src_dst.push((file.to_path_buf(), obj.clone()));
+                fs::create_dir_all(&obj.parent().unwrap()).unwrap();
+            }
+
             objects.push(obj);
         }
-        self.compile_objects(&src_dst);
-        self.assemble(lib_name, &dst.join(output), &objects);
+
+        // if objects are queued to build
+        if src_dst.len() > 0 {
+            self.compile_objects(&src_dst);
+            self.assemble(lib_name, &dst.join(output), &objects);
```

Comment:
We may want to also check whether the output archive exists because otherwise it could be a case where a build was interrupted with all object files but the output wasn't created yet.